File: src/sr-inference/parsing.jl
module ParsingModule

import SymbolicRegression: Node
using Distributions: Categorical

# Define grammar rules similar to Python version
grammar_str = """
S -> 'ADD' S S | 'SUB' S S | 'MUL' S S | 'DIV' S S
S -> 'SIN' S | 'COS' S | 'EXP' S | 'TANH' S | 'COSH' S | 'SINH' S 
S -> 'CON'
S -> 'x1'
END -> 'END'
"""

const OPERATOR_ARITY = Dict{String, Int}(
    # Elementary functions
    "ADD" => 2,
    "SUB" => 2,
    "MUL" => 2,
    "DIV" => 2,

    # Trigonometric Functions
    "SIN" => 1,
    "COS" => 1,
    "EXP" => 1,

    # Hyperbolic Functions
    "SINH" => 1,
    "COSH" => 1,
    "TANH" => 1,

    # FIXME: Handle this differently!
    "x1" => 0
)

# Split each production rule into separate lines
grammar_lines = String[]
for line in split(grammar_str, "\n")
    line = strip(line)
    if !isempty(line)
        lhs, rhs = split(line, "->")
        lhs = strip(lhs)
        # Split on | and create new lines
        for rule in split(rhs, "|")
            push!(grammar_lines, "$lhs -> $(strip(rule))")
        end
    end
end
grammar_str = join(grammar_lines, "\n")

function _create_grammar_masks(grammar_str::String)
    # Collect all LHS symbols and unique set
    all_lhs = String[]
    unique_lhs = String[]
    
    # Parse grammar string to get productions
    for line in split(grammar_str, "\n")
        if !isempty(strip(line))
            lhs = strip(split(line, "->")[1])
            push!(all_lhs, lhs)
            if !(lhs in unique_lhs)
                push!(unique_lhs, lhs)
            end
        end
    end

    # Create masks matrix - each row corresponds to a unique LHS symbol
    # and has 1s for productions with that LHS
    masks = falses(length(unique_lhs), length(all_lhs))
    for (i, symbol) in enumerate(unique_lhs)
        for (j, lhs) in enumerate(all_lhs)
            masks[i,j] = (symbol == lhs)
        end
    end

    allowed_prod_idx = findall(vec(any(masks, dims=1)))

    return masks, allowed_prod_idx, unique_lhs
end
masks, allowed_prod_idx, unique_lhs = _create_grammar_masks(grammar_str)

mutable struct nn_config
    nbin::Int
    nuna::Int
    nvar::Int

    seq_len::Int
    nb_onehot_cats::Int

    function nn_config(;nbin::Int, nuna::Int, nvar::Int, seq_len::Int)
        nb_onehot_cats = nbin + nuna + nvar + 2  # +1 for constants, +1 for end token
        return new(nbin, nuna, nvar, seq_len, nb_onehot_cats)
    end
end


function node_to_onehot(node::Node{T}, cfg::nn_config)::Tuple{Bool, Matrix{Float32}} where T <: Number
    prefix = _tree_to_prefix(node)
    idx = [_node_to_token_idx(node, cfg) for node in prefix]
    success, onehot, consts = _onehot_encode(idx, cfg)
    ~success && return false, nothing

    x = zeros(Float32, cfg.seq_len, cfg.nb_onehot_cats+1)
    x[:, 1:cfg.nb_onehot_cats] = onehot
    x[:, cfg.nb_onehot_cats+1] = consts
    return true, x
end

function _onehot_encode(idx::Vector{Tuple{Int, Float64}}, cfg::nn_config)::Tuple{Bool, BitMatrix, Vector{Float64}}
    onehot = falses(cfg.seq_len, cfg.nb_onehot_cats)
    consts = zeros(Float64, cfg.seq_len)

    if length(idx) > cfg.seq_len
        return false, nothing, nothing
    end
    for (node_i, (token_idx, token_val)) in enumerate(idx)  # Each node has corresponding index and value
        onehot[node_i, token_idx] = true
        consts[node_i] = token_val
    end
    # Pad with empty token (last category)
    onehot[length(idx)+1:end, end] .= true 
    return true, onehot, consts
end

function _tree_to_prefix(tree::Node{T})::Vector{Node{T}} where T <: Number
    result = [tree]
    if tree.degree >= 1
        append!(result, _tree_to_prefix(tree.l))
    end
    if tree.degree == 2
        append!(result, _tree_to_prefix(tree.r))
    end
    return result
end

function _node_to_token_idx(node::Node{T}, cfg::nn_config)::Tuple{Int, Float64} where T <: Number
    offset_unaop = cfg.nbin
    offset_const = offset_unaop + cfg.nuna
    offset_var = offset_const + cfg.nvar
    if node.degree == 2
        return (node.op, 0)
    elseif node.degree == 1
        return (offset_unaop + node.op, 0)
    elseif node.degree == 0
        if node.constant
            return (offset_const + 1, node.val)  # Only 1 const token
        else
            return (offset_var + node.feature, 0)
        end
    end
end


function logits_to_prods(logits::Matrix{Float32}, sample::Bool=false, max_length::Int=15)::Vector{Tuple{String, String}}
    # Initialize empty stack with start symbol 'S'
    stack = ["S"]
    
    # Split logits into productions and constants
    logits_prods = logits[:, 1:end-1] 
    constants = logits[:, end]
    
    prods = []
    t = 1
    
    while !isempty(stack)
        alpha = pop!(stack)  # Current LHS toke
        
        # Get mask for current symbol
        symbol_idx = findfirst(==(alpha), unique_lhs)
        mask = masks[symbol_idx, :]
        
        # Calculate probabilities
        probs = mask .* exp.(logits_prods[t, :])
        tot = sum(probs)
        @assert tot > 0 "Sum of probs is 0 at t=$t. Probably due to bad mask or invalid logits?"
        probs = probs ./ tot
        
        # Select production rule
        if sample
            i = rand(Categorical(probs))
        else
            _, i = findmax(probs)
        end
        
        # Get selected rule
        rule = split(grammar_str, "\n")[i]
        lhs, rhs = split(strip(rule), "->")
        lhs = strip(lhs)
        rhs = strip(rhs)
        
        # If rule produces CONST, replace with actual constant
        if rhs == "'CON'"
            rhs = string(constants[t])
        end
        
        # Add production to list
        push!(prods, (lhs, rhs))
        
        # Add RHS nonterminals to stack in reverse order
        rhs_symbols = split(rhs)
        for symbol in reverse(rhs_symbols)
            clean_symbol = replace(symbol, "'" => "")
            if clean_symbol in unique_lhs
                push!(stack, clean_symbol)
            end
        end
        
        t += 1
        if t > max_length
            break
        end
    end
    
    return prods
end

function prods_to_tree(prods::Vector{Tuple{String, String}}, OP_INDEX::Dict{String, Int}, feature::Int)
    # global prods stack
    # Create node for each production
    # Depending on arity, create 0, 1 or 2 children
    # 
    prefix_list = _prods_to_prefix(prods, OP_INDEX, feature)
    tree = _prefix_to_tree!(prefix_list)

    return tree
end

function _prods_to_prefix(prods::Vector{Tuple{String, String}}, OP_INDEX::Dict{String, Int}, feature::Int)::Vector{Node{Float64}}
    prefix_list = []
    for prod in prods
        op_match = match(r"'([^']+)'", prod[2])  # Alternatively, use prod to infer arity?
        if op_match !== nothing
            op = op_match.captures[1]
            arity = OPERATOR_ARITY[op]
            if arity == 0
                push!(prefix_list, Node{Float64}(; feature=feature))  # FIXME: Only univariate for now
            elseif arity == 1
                push!(prefix_list, _make_childless_op(arity, OP_INDEX[op], Float64))
            elseif arity == 2
                push!(prefix_list, _make_childless_op(arity, OP_INDEX[op], Float64))
            end
        else  # Constant
            push!(prefix_list, Node{Float64}(; val=parse(Float64, prod[2])))
        end
    end
    return prefix_list
end


"""
Copied from datagen/ExpressionGenerator.jl. Consolidate.
"""
function _make_childless_op(degree::Int, op_index::Int, ::Type{T})::Node{T} where {T<:Number}
    op_node = Node{T}()
    op_node.degree = degree
    op_node.op = op_index
    return op_node
end

"""
Copied from datagen/ExpressionGenerator.jl. Consolidate.
"""
function _prefix_to_tree!(prefix_list::Vector{Node{T}})::Node{T} where {T<:Number}
    function build_subtree()
        if isempty(prefix_list)
            error("Unexpected end of prefix list")
        end
        node = popfirst!(prefix_list)
        if node.degree == 0
            # Leaf node, no children to add
        elseif node.degree == 1
            node.l = build_subtree()
        elseif node.degree == 2
            node.l = build_subtree()
            node.r = build_subtree()
        else
            error("Invalid node degree: $(node.degree)")
        end
        return node
    end

    if isempty(prefix_list)
        error("Empty prefix list")
    end
    return build_subtree()
end

"""
Selects a subtree to use for neural sampling. Requirements for the subtree:
    - Univariate (only one feature)
    - 2-14 nodes

Returns the subtree and the feature used in the subtree.

FIXME: Make stochastic.
"""
function select_subtree(t::Node)::Tuple{Node, Node, Int}
    node_list = _tree_to_prefix(t)

    valid_subtrees = []
    for node in node_list
        subtree_nodes = _tree_to_prefix(node)
        
        # Check size constraint
        if length(subtree_nodes) < 2 || length(subtree_nodes) > 14
            continue
        end
        
        # Track which features are used
        features_used = Set{Int}()
        for n in subtree_nodes
            if n.degree == 0 && !n.constant
                push!(features_used, n.feature)
            end
        end
        
        # Check univariate constraint
        if length(features_used) == 1
            # Find parent node
            parent = nothing
            for potential_parent in node_list
                if potential_parent.l === node || potential_parent.r === node
                    parent = potential_parent
                    break
                end
            end
            push!(valid_subtrees, (node, parent, first(features_used)))
        end
    end

    if isempty(valid_subtrees)
        return nothing
    end

    return valid_subtrees[rand(1:length(valid_subtrees))]
    
end

end

File: src/utils/parsing.py
import torch
import numpy as np
from nltk import Nonterminal
from nltk.grammar import Production
from torch.distributions import Categorical
from typing import List, Tuple, Literal
import sympy as sp
from sympy.utilities.lambdify import lambdify

from src.utils.grammar import get_mask, NT, GCFG

CUSTOM_FUNCTIONS = {
    'zero_sqrt': lambda x: np.sqrt(np.maximum(x, 0)),  # Safe sqrt that returns 0 for negative inputs
}

OPERATOR_ARITY = {
    # Elementary functions
    "ADD": 2,
    "SUB": 2,
    "MUL": 2,
    "DIV": 2,
    "POW": 2,
    "INV": 1,
    "SQRT": 1,
    "ZERO_SQRT": 1,  # Custom safe sqrt
    "EXP": 1,
    "LN": 1,
    "ABS": 1,

    # Trigonometric Functions
    "SIN": 1,
    "COS": 1,
    "TAN": 1,

    # Trigonometric Inverses
    "ASIN": 1,
    "ACOS": 1,
    "ATAN": 1,

    # Hyperbolic Functions
    "SINH": 1,
    "COSH": 1,
    "TANH": 1,
    "COTH": 1,
}

OPERATORS = OPERATOR_ARITY.keys()
UNARY_OP_TOKENS  = [k for k, v in OPERATOR_ARITY.items() if v == 1]

def prods_to_prefix(prods, to_string=False, require_no_nonterminals=True):
    """
    Takes a list of productions and a list of constants and returns a string representation of the equation.
    """
    seq = [prods[0].lhs()]  # Start with LHS of first rule (always nonterminal start)
    for prod in prods:
        if str(prod.lhs()) == 'Nothing':  # Padding rule. Reached end.
            break
        for ix, s in enumerate(seq):  # Applying rule to first element in seq that matches lhs
            if s == prod.lhs():
                seq = seq[:ix] + list(prod.rhs()) + seq[ix+1:]  # Replace LHS with RHS
                break

    if to_string:
        try:
            return ' '.join(seq)
        except TypeError:
            print(f'Error. Could not create equation from {seq = }; {prods = }')
            return ''
    else:
        if require_no_nonterminals:
            assert not any(isinstance(s, Nonterminal) for s in seq), f'Found nonterminals in sequence: {seq}'
        return seq

def logits_to_prods(
    logits, 
    start_symbol: Nonterminal = NT, 
    sample=False, 
    max_length=15, 
    insert_const=True, 
    const_token='CON', 
    replace_const: Literal['numerical', 'placeholder', 'nothing', 'numerical_rounded'] = 'numerical', 
    round_const_decimals=2
) -> List[Production]:
    stack = [start_symbol]

    logits_prods = logits[:, :-1]
    constants = logits[:, -1]

    prods = []
    t = 0  # "Time step" in sequence
    j = 0  # Index of constant
    while len(stack) > 0:
        alpha = stack.pop()  # Alpha is notation in paper: current LHS token
        mask = get_mask(alpha)
        probs = mask * logits_prods[t].exp()
        assert (tot := probs.sum()) > 0, f"Sum of probs is 0 at t={t}. Probably due to bad mask or invalid logits?"
        probs = probs / tot

        if sample:
            m = Categorical(probs)
            i = m.sample()
        else:
            _, i = probs.max(-1) # argmax

        # select rule i
        rule = GCFG.productions()[i.item()]

        # If rule has -> [CONST] add const
        if insert_const and (rule.rhs()[0] == const_token):
            if replace_const == 'numerical':
                rule = Production(lhs=rule.lhs(), rhs=(str(constants[t].item()),))
            elif replace_const == 'numerical_rounded':
                rule = Production(lhs=rule.lhs(), rhs=(str(round(constants[t].item(), round_const_decimals)),))
            elif replace_const == 'placeholder':
                placeholder_name = f'CON_{j}'
                rule = Production(lhs=rule.lhs(), rhs=(placeholder_name,))
                j += 1
            elif replace_const == 'nothing':
                pass
            else:
                raise ValueError(f'Unknown replace_const mode: {replace_const}')

        prods.append(rule)
        # add rhs nonterminals to stack in reversed order
        for symbol in reversed(rule.rhs()):
            if isinstance(symbol, Nonterminal):
                stack.append(symbol)
        t += 1
        if t == max_length:
            break
    return prods

def logits_to_prefix(logits, syntax_cats: list[str], sample=False, max_length=15):
    print('Warning: Not using constants in syntax probs.')

    consts = logits[:, -1]  # FIXME: Replace const placeholders
    syntax = logits[:, :-1]

    token_idx = []
    probs = torch.softmax(syntax, dim=-1)  # Convert logits to probabilities, excluding the last column (constants)
    
    for t in range(max_length):
        if sample:
            dist = torch.distributions.Categorical(probs[t])
            idx = dist.sample()
        else:
            idx = torch.argmax(probs[t])
        
        token_idx.append(idx.item())
        
        # Break if we've reached the end of the sequence
        if idx == (len(syntax_cats) - 1):  # Assuming the last category is an end token
            break
    
    # Convert token indices to actual tokens
    tokens = [syntax_cats[idx] for idx in token_idx]
    
    return tokens

def _is_leaf_token(token: str) -> bool:
    """
    A leaf node does not need paranthesis around it.

    TODO: Check if this are all the cases that need parenthesis.
    """
    return not any(op in token for op in ['*', '+', '-', '/', '**'])

def _write_infix(token: str, args: List[str]) -> str:
    """
    Convert prefix operator with arguments to a infix string that SymPy can parse.
    FIXME: Remove unnecessary parenthesis in case argument is already enclosed by unary operator.
    """
    
    if token in UNARY_OP_TOKENS:  # FIXME: Check parenthesis
        return f"{token}({args[0]})"
    else:
        args = ['(' + arg + ')' if not _is_leaf_token(arg) else arg for arg in args]
        if token == "ADD":
            return f"{args[0]}+{args[1]}"
        elif token == "SUB":
            return f"{args[0]}-{args[1]}"
        elif token == "MUL":
            return f"{args[0]}*{args[1]}"
        elif token == "DIV":
            return f"{args[0]}/{args[1]}"
        elif token == "POW":
            return f"{args[0]}**{args[1]}"
        elif token == "ABS":
            return f"Abs({args[0]})"
        elif token == "INV":
            return f"1/{args[0]}"
        else:
            raise Exception(f"Unknown token in prefix expression: {token}, with arguments {args}")

def _prefix_to_infix(expr: List[str], variables=None) -> Tuple[str, List[str]]:
    """
    Parse an expression in prefix mode, and output it in either:
        - infix mode (returns human readable string)
        - develop mode (returns a dictionary with the simplified expression)
    """
    if len(expr) == 0:
        raise Exception("Empty prefix list.")
    t = expr[0]
    if t in OPERATORS:
        args = []
        l1 = expr[1:]
        for _ in range(OPERATOR_ARITY[t]):
            i1, l1 = _prefix_to_infix(l1, variables)
            args.append(i1)
        return _write_infix(t, args), l1
    elif t in variables:
        return t, expr[1:]
    else: # Constant
        val = expr[0]
        return str(val), expr[1:]
    
def prefix_to_infix(expr: List[str], variables=['x1']) -> List[str]:
    infix_str, remainder = _prefix_to_infix(expr, variables)
    if len(remainder) > 0:
        raise Exception(f'Invalid prefix expression "{expr}". Successfully parsed "{infix_str}" but "{remainder}" is still remaining.')
    return infix_str

def logits_to_infix(logits, sample=False, replace_const='numerical', round_const_decimals=2):
    # FIXME: Add variables, GCFG pass-through

    assert len(logits.shape) == 2, "Logits should be 2D, no batch dimension"
    prods = logits_to_prods(logits, sample=sample, replace_const=replace_const, round_const_decimals=round_const_decimals)
    prefix = prods_to_prefix(prods, require_no_nonterminals=True)
    infix = prefix_to_infix(prefix, variables=['x1'])
    return infix

def eval_from_logits(logits, val_x):
    assert len(val_x.shape) == 1, "val_x should be 1D"
    x1 = sp.Symbol('x1')

    infix = logits_to_infix(logits)
    expr = sp.sympify(infix.lower())

    # Pass custom_functions as third argument
    # Check for infinity or similar in expression
    assert not expr.has(sp.zoo), f"Expression {expr} contains infinity"
    func = lambdify(x1, expr, ('numpy', CUSTOM_FUNCTIONS))

    with np.errstate(all='ignore'):
        val_y = func(val_x)

    if not isinstance(val_y, np.ndarray):
        # sp.simpify simplifies expressions which may result in `expr` being a constant. 
        # In this case calling func() simply returns this constant instead of a full numpy array
        val_y = np.repeat(val_y, val_x.shape[0])

    return val_y
